lib/tree.rb

```diff
@@ -16,40 +16,122 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(log n)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+
+    if @root == nil
+      @root = new_node
+    else
+      add_helper(@root, new_node)
+    end
+
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current, new_node)
+    if new_node.key <= current.key
+      if current.left.nil?
+        return current.left = new_node
+      else
+        add_helper(current.left, new_node)
+      end
+    else
+      if current.right.nil?
+        return current.right = new_node
+      else
+        add_helper(current.right, new_node)
+      end
+    end
+  end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(log n)
   def find(key)
```

Comment:
👍 Since you are using a loop the space complexity is actually O(1) since you are not building a system stack.